### ISSUE TITLE
Optionally exclude node_modules and bower_components folder

### ISSUE THREAD

**Issue by Turbo87** [Turbo87/intellij-emberjs#24]
As suggested in #22 we should add a plugin-wide option whether the `node_modules` and `bower_components` folder should be marked as library folders or instead excluded from indexing at all. I would personally use the first option because once the initial index is done it usually stays quite fast but I can see that people might disagree.

**Comment by orf** [comment]
It's a hard one, I'm in favour of keeping it, but if you have a sluggish machine then indexing can/will freeze everything. You need to specify this while creating the project I think, you can't really alter the preference while the IDE is frozen :)

Perhaps when creating the project we can check the number of folders in `node_modules`, and if there are more than X display a box saying "This may take a while to index, do you want to continue or ignore?".

**Comment by dfreeman** [comment]
@orf It's not just if you have a sluggish machine – once your apps get large enough to pull in enough dependencies, it's going to get bad no matter how snappy your hardware is :)

I also suspect that the general issues with `fseventsd` not getting along well with Ember CLI on OS X makes the indexing issues worse for Mac users as well, because there's so much thrashing in general within the project directory.

Honestly, I'd settle for just being able to manually remove `node_modules` and `bower_components` as libraries and have them stay excluded – currently they reappear as library directories every time I restart the IDE.

**Comment by Turbo87** [comment]
@dfreeman I'm using a MacBook Air for development at the moment and don't have much issues with that. Once the initial indexing is done the IDE is usually quite fast again.

As mentioned before: I'm open to pull requests on this "feature", but it is low priority for me personally.

**Comment by runspired** [comment]
`node_modules` and `bower_components` should absolutely never be indexed for Web projects, they are highly volatile.  I had to disable this plugin because it was remarking them as libraries every time I opened a project even though they were marked as excluded and I'd removed the library mapping.

**Comment by runspired** [comment]
On average, I `rm -rf node_modules && npm cache clean && npm install` at least 10x a day, it takes Webstorm 15min+ of CPU cooking to finish indexing the >100mb of files that results in.

**Comment by Kerrick** [comment]
Even marking the folders as excluded from the preferences -> project -> directories seems to be overridden by the plugin. Because I've got so many dependencies, the plugin slows my machine to an unusable crawl. I'd love to use the plugin, though!

**Comment by runspired** [comment]
@Kerrick This will help, it doesn't affect WebStorm's indexing but should stop Spotlight from also indexing: https://gist.github.com/runspired/b5552be57d3c8a5a7821

**Comment by runspired** [comment]
@Turbo87 this appears to be what we need to implement: http://www.jetbrains.org/intellij/sdk/docs/basics/getting_started/creating_an_action.html

**Comment by Turbo87** [comment]
@runspired I'm not sure how an "Action" is supposed to help here.

What we need to do is add `addExcludeFolder()` calls for `node_modules` and `bower_components` to https://github.com/Turbo87/intellij-emberjs/blob/master/src/main/kotlin/com/emberjs/project/EmberProjectComponent.kt#L130. As I've mentioned before though this should be configurable so the real work here is finding out how to add a configuration panel to the IDE settings.

**Comment by runspired** [comment]
@Turbo87 that actions api is from what I can tell how you add settings to the preferences panel as well.  May have misread something.  Also note, you can't BOTH `addExcludeFolder()` and `addLibraryFolder()`, the library definition always trumps and causes indexing.

**Comment by runspired** [comment]
Appears this is what we actually need to use: https://confluence.jetbrains.com/display/IDEADEV/Customizing+the+IDEA+Settings+Dialog

**Comment by Ramblurr** [comment]
This issue rears its ugly head especially when you are using `npm link` to  link a locally-devloped addon into your main app's node_modules folder. Unlike other items in node_modules, these linked addons include their own node_module folder (e.g., app-repo/node_modules/local-linked-addon/node_modules/....) which makes the indexing process take a very, very long time.

**Comment by k3n** [comment]
Had to disable the plugin due to this issue -- I'm in the same boat as @runspired, and echo his first 2 comments on this issue (from 2/17). I also echo @Ramblurr's comment re: `npm link` on a locally-dev'd addon, which I also do.

**Comment by Ramblurr** [comment]
I temporarily worked around this by building the plugin with this patch

```
diff --git a/src/main/kotlin/com/emberjs/project/EmberProjectComponent.kt b/src/main/kotlin/com/emberjs/project/EmberProjectComponent.kt
index e32b780..2dab046 100644
--- a/src/main/kotlin/com/emberjs/project/EmberProjectComponent.kt
+++ b/src/main/kotlin/com/emberjs/project/EmberProjectComponent.kt
@@ -64,7 +64,7 @@ class EmberProjectComponent(val project: Project) : AbstractProjectComponent(pro
             ApplicationManager.getApplication().invokeLater {
                 ApplicationManager.getApplication().runWriteAction {
                     roots.forEach {
-                        createLibrary("node_modules", project, it)
+                        // createLibrary("node_modules", project, it)
                         createLibrary("bower_components", project, it)
                     }

@@ -137,6 +137,7 @@ class EmberProjectComponent(val project: Project) : AbstractProjectComponent(pro
         entry.addSourceFolder("$rootUrl/tests/integration", TEST_SOURCE)
         entry.addExcludeFolder("$rootUrl/dist")
         entry.addExcludeFolder("$rootUrl/tmp")
+        entry.addExcludeFolder("$rootUrl/node_modules")
     }

     private val VirtualFile.isEmberFolder: Boolean
```
